numpy/core/src/multiarray/ctors.c

```diff
@@ -1076,11 +1076,32 @@ PyArray_NewLikeArrayWithShape(PyArrayObject *prototype, NPY_ORDER order,
     }
 
     /* Logic shared by `empty`, `empty_like`, and `ndarray.__new__` */
-    if (PyDataType_REFCHK(PyArray_DESCR((PyArrayObject *)ret))) {
-        PyArray_FillObjectArray((PyArrayObject *)ret, Py_None);
-        if (PyErr_Occurred()) {
-            Py_DECREF(ret);
-            return NULL;
+    PyArray_Descr* descr = PyArray_DESCR((PyArrayObject *)ret);
+    if (PyDataType_REFCHK(descr)) {
+        if (NPY_DT_is_legacy(NPY_DTYPE(dtype))) {
+            PyArray_FillObjectArray((PyArrayObject *)ret, Py_None);
+            if (PyErr_Occurred()) {
+                Py_DECREF(ret);
+                return NULL;
+            }
+        }
+        else {
+            // Currently we assume all non-legacy dtypes that have with the
+            // NPY_ITEM_REFCOUNT flag either represent heap-allocated dtypes or
+            // represent python objects. In the latter case the dtype is
+            // responsible for managing initialization and reference counts. In
+            // both cases initializing to NULL makes sense.
+            //
+            // In the future we might adjust this and have separate logic for
+            // new dtypes that hold heap-allocated data and new dtypes that hold
+            // python objects, particularly if we want to allow releasing the
+            // GIL in the former case.
+            char *optr = PyArray_DATA((PyArrayObject*)ret);
+            npy_intp n = PyArray_SIZE((PyArrayObject*)ret);
+            for (npy_intp i = 0; i < n; i++) {
+                optr = NULL;
+                optr += descr->elsize;
+            }
```

Comment:
OK, so I don't think this is better unfortunately.  We already guarantee that the full memory is NULLed right now, so you might as well do nothing if you do this.

Now, `PyArray_FillObjectArray` is also used in two other use cases:
1. for `np.zero` (and `arr.imag`) to fill with 0 rather than None
2. For structured dtypes and you string could be included (hitting the first branch and then failing the assert)

It is public API also, but lets ignore that hideous little detail, since it seems a weird thing to use by downstream (especially on arrays that you didn't explicitly create with `dtype=object`).

---

Now, for the `None` filling, we really know what we need to do: Nothing at all is fine for *all* non-legacy dtypes.

For the zero filling (or all other values), that could be incorrect and actually filling with zero might be better.  *except* that it is wrong for strings, because `np.zeros(4, dtype=StringDType())` shouldn't give you a lot of `"0"` strings :).

My feeling, maybe we should cut our losses?  `PyArray_FillObjectArray` is broken, and for non-legacy dtypes maybe it might as well do *nothing* in all branches.  That will allow us to keep using it indefinitely internally for `np.empty()`.

For `np.zeros()` (and imag), I suspect we need to find a new solution anyway, something like `dtype.give_me_a_zero(buffer)` (the same way I am now handling the identity values).
We could do `dtype.give_me_an_uninitialized()` also.  But for now only `object` dtype is special (we need to explicitly `None` fill, although I fixed cython, so it may be becoming a non-issue in practice.)

Reply:
Again, thanks for your careful thought here. Let's leave adding a new hook for supplying zero and empty initial values to a future PR. For now I've added an early return from `PyArray_FillObjectArray` for non-legacy dtypes, which I've confirmed works fine.